File: frontend/src/types/typesNewTask.ts
import { z } from "zod"

export type TypeThemes = {
    id: number | null
    name: string
    description?: string
    match_percentage: number
    reason: string
}

export type TypeSubTask = {
    title: string
    description: string
    instruction: string
    continuance: number
    motivation: string
}

export type TypeAssociation = {
    id: number
    name: string
    description: string
    reason:string
}

export type TypeNewTask = {
	title: string
	description: string
    motivation?: string
	match_themes?: TypeThemes[]
    new_themes?: TypeThemes[]
    subtasks?: TypeSubTask[]
    risk?: 0 | 1 | 2 | 3
    risk_explanation?: string
    risk_proposals?: string
    impact?: 0 | 1 | 2 | 3
    states?: {
        physical: TypeAssociation[]
        intellectual: TypeAssociation[]
        emotional: TypeAssociation[]
        motivational: TypeAssociation[]
        social: TypeAssociation[]
    }
    action_type?: TypeAssociation[]
    stress?: TypeAssociation[]
    energy_level?: TypeAssociation[]
    deadline?: string | null     // время дедлайна
    activation?: string | null   // время активации задачи
    taskchecks?: string[]        // даты проверки задачи
}


export const newTaskSchema = z.object({
    title: z.string().min(6, "Task title must be at least 6 characters"),
    description: z.string(),
    motivation: z.string().optional(),
    match_themes: z.array(
        z.object({
            id: z.number().nullable(),
            name: z.string(),
            description: z.string().optional(),
            match_percentage: z.number(),
            reason: z.string()
        })
    ).optional(),
    new_themes: z.array(
        z.object({
            id: z.number().nullable(),
            name: z.string(),
            description: z.string().optional(),
            match_percentage: z.number(),
            reason: z.string()
        })
    ).optional(),
    subtasks: z.array(
        z.object({
            title: z.string(),
            description: z.string(),
            instruction: z.string(),
            continuance: z.number(),
            motivation: z.string()
        })
    ).optional(),
    risk: z.union([z.literal(0), z.literal(1), z.literal(2), z.literal(3)]).optional(),
    risk_explanation: z.string().optional(),
    risk_proposals: z.string().optional(),
    impact: z.union([z.literal(0), z.literal(1), z.literal(2), z.literal(3)]).optional(),
    states: z.object({
        physical: z.array(
            z.object({
                id: z.number(),
                name: z.string(),
                description: z.string(),
                reason: z.string()
            })
        ),
        intellectual: z.array(
            z.object({
                id: z.number(),
                name: z.string(),
                description: z.string(),
                reason: z.string()
            })
        ),
        emotional: z.array(
            z.object({
                id: z.number(),
                name: z.string(),
                description: z.string(),
                reason: z.string()
            })
        ),
        motivational: z.array(
            z.object({
                id: z.number(),
                name: z.string(),
                description: z.string(),
                reason: z.string()
            })
        ),
        social: z.array(
            z.object({
                id: z.number(),
                name: z.string(),
                description: z.string(),
                reason: z.string()
            })
        )
    }).optional(),
    action_type: z.array(
        z.object({
            id: z.number(),
            name: z.string(),
            description: z.string(),
            reason: z.string()
        })
    ).optional(),
    stress: z.array(
        z.object({
            id: z.number(),
            name: z.string(),
            description: z.string(),
            reason: z.string()
        })
    ).optional(),
    energy_level: z.array(
        z.object({
            id: z.number(),
            name: z.string(),
            description: z.string(),
            reason: z.string()
        })
    ).optional(),
    deadline: z.string().nullable().optional(),
    activation: z.string().nullable().optional(),
    taskchecks: z.array(z.string()).optional()
});

File: backend/schemas/types_new_task.py
from typing import List, Optional, Literal
from pydantic import BaseModel

class TaskGenerateRequest(BaseModel):
    text: str
    description: Optional[str] = ""

class TypeThemes(BaseModel):
    id: Optional[int] = None
    name: str
    description: Optional[str] = None
    match_percentage: int
    reason: str

class TypeSubTask(BaseModel):
    title: str
    description: str
    instruction: str
    continuance: int
    motivation: str

class TypeAssociation(BaseModel):
    id: int
    title: str
    description: str
    reason: str

class RiskExplanation(BaseModel):
    reason: str
    proposals: str

class States(BaseModel):
    physical: List[TypeAssociation]
    intellectual: List[TypeAssociation]
    emotional: List[TypeAssociation]
    motivational: List[TypeAssociation]
    social: List[TypeAssociation]

class NewTaskRequest(BaseModel):
    title: str
    description: str
    motivation: Optional[str] = None
    match_themes: Optional[List[TypeThemes]] = None
    new_themes: Optional[List[TypeThemes]] = None
    subtasks: Optional[List[TypeSubTask]] = None
    risk: Optional[Literal[0, 1, 2, 3]] = None
    risk_explanation: Optional[RiskExplanation] = None
    impact: Optional[Literal[0, 1, 2, 3]] = None
    states: Optional[States] = None
    action_type: Optional[List[TypeAssociation]] = None
    stress: Optional[List[TypeAssociation]] = None
    energy_level: Optional[List[TypeAssociation]] = None
    deadline: Optional[str] = None
    activation: Optional[str] = None
    taskchecks: Optional[List[str]] = None
    
